File: Frontend/src/oral-interview/constants/oralTiming.ts
/**
 * Timing strict entretien oral (aligné avec le backend `app.constants.oral_timing`).
 */

export const ORAL_PREP_TIME = 7;

export const ORAL_ANSWER_TIME_DEFAULT = 60;

export const ORAL_ANSWER_MARGIN_SECONDS = 5;

/** Tolérance réseau / encodage côté serveur */
export const ORAL_DURATION_REJECT_MARGIN = ORAL_ANSWER_MARGIN_SECONDS;

export const QUESTION_TIME_MAP = {
  intro: 45,
  normal: 60,
  advanced: 75,
} as const;

/** Seuil visuel : derniers 10 s */
export const ANSWER_WARN_LAST_SECONDS = 10;
/** Seuil visuel : derniers 5 s (rouge) */
export const ANSWER_CRITICAL_LAST_SECONDS = 5;

/** Aligné sur `backend/app/constants/oral_timing.py` (mêmes sous-chaînes). */
const ADVANCED_SNIPPETS = [
  "stratég",
  "strateg",
  "architecture",
  "due diligence",
  "restructuration",
  "transformation majeure",
  "m&a",
  "maîtrise",
  "pilotage",
  "vision",
] as const;

/**
 * Les 3 premières questions correspondent aux questions fixes d’introduction (banque orale).
 */
export function getAnswerSecondsForQuestion(questionText: string, questionIndexZeroBased: number): number {
  if (questionIndexZeroBased < 3) {
    return QUESTION_TIME_MAP.intro;
  }
  const t = (questionText || "").toLowerCase();
  if (ADVANCED_SNIPPETS.some((x) => t.includes(x))) {
    return QUESTION_TIME_MAP.advanced;
  }
  return QUESTION_TIME_MAP.normal;
}

export function getMaxAllowedClientSeconds(allowedAnswerSeconds: number): number {
  return allowedAnswerSeconds + ORAL_ANSWER_MARGIN_SECONDS;
}

File: backend/app/constants/oral_timing.py
"""
Durées autorisées pour les réponses orales (cohérent avec `Frontend/src/oral-interview/constants/oralTiming.ts`).
"""

ORAL_PREP_TIME = 7
ORAL_ANSWER_TIME_DEFAULT = 60
ORAL_ANSWER_MARGIN_SECONDS = 5

QUESTION_TIME_INTRO = 45
QUESTION_TIME_NORMAL = 60
QUESTION_TIME_ADVANCED = 75


def max_answer_seconds_for_question(question_order_1based: int, question_text: str) -> int:
    """
    `question_order` : 1-based (aligné API).
    Intro : ordres 1–3. Sinon heuristique « avancé » sur le libellé.
    """
    idx0 = int(question_order_1based) - 1
    if idx0 < 3:
        return QUESTION_TIME_INTRO
    t = (question_text or "").lower()
    if any(
        x in t
        for x in (
            "stratég",
            "strateg",
            "architecture",
            "due diligence",
            "restructuration",
            "transformation majeure",
            "m&a",
            "maîtrise",
            "pilotage",
            "vision",
        )
    ):
        return QUESTION_TIME_ADVANCED
    return QUESTION_TIME_NORMAL


def max_answer_seconds_with_margin(question_order_1based: int, question_text: str) -> int:
    return max_answer_seconds_for_question(question_order_1based, question_text) + ORAL_ANSWER_MARGIN_SECONDS
